sympy/matrices/common.py

```diff
@@ -2500,3 +2499,219 @@ def classof(A, B):
+    with their diagonal on the indicated diagonal:
+
+    >>> vert = [1, 2, 3]
+    >>> banded({0: vert}, cols=3)
+    Matrix([
+    [1, 0, 0],
+    [2, 1, 0],
+    [3, 2, 1],
+    [0, 3, 2],
+    [0, 0, 3]])
+
+    >>> banded(4, {0: ones(2)})
+    Matrix([
+    [1, 1, 0, 0],
+    [1, 1, 0, 0],
+    [0, 0, 1, 1],
+    [0, 0, 1, 1]])
+
+    Errors are raised if the designated size will not hold
+    all values an integral number of times. Here, the rows
+    are designated as odd (but an even number is required to
+    hold the off-diagonal 2x2 ones):
+
+    >>> banded({0: 2, 1: ones(2)}, rows=5)
+    Traceback (most recent call last):
+    ...
+    ValueError:
+    sequence does not fit an integral number of times in the matrix
+
+    And here, an even number of rows is given...but the square
+    matrix has an even number of columns, too. As we saw
+    in the previous example, an odd number is required:
+
+    >>> banded(4, {0: 2, 1: ones(2)})  # trying to make 4x4 and cols must be odd
+    Traceback (most recent call last):
+    ...
+    ValueError:
+    sequence does not fit an integral number of times in the matrix
+
+    A way around having to count rows is to enclosing matrix elements
+    in a tuple and indicate the desired number of them to the right:
+
+    >>> banded({0: 2, 2: (ones(2),)*3})
+    Matrix([
+    [2, 0, 1, 1, 0, 0, 0, 0],
+    [0, 2, 1, 1, 0, 0, 0, 0],
+    [0, 0, 2, 0, 1, 1, 0, 0],
+    [0, 0, 0, 2, 1, 1, 0, 0],
+    [0, 0, 0, 0, 2, 0, 1, 1],
+    [0, 0, 0, 0, 0, 2, 1, 1]])
+
+    An error will be raised if more than one value
+    is written to a given entry. Here, the ones overlap
+    with the main diagonal if they are placed on the
+    first diagonal:
+
+    >>> banded({0: (2,)*5, 1: (ones(2),)*3})
+    Traceback (most recent call last):
+    ...
+    ValueError: collision at (1, 1)
+
+    By placing a 0 at the bottom left of the 2x2 matrix of
+    ones, the collision is avoided:
+
+    >>> u2 = Matrix([
+    ... [1, 1],
+    ... [0, 1]])
+    >>> banded({0: (2,)*5, 1: (u2,)*3})
+    Matrix([
+    [2, 1, 1, 0, 0, 0, 0],
+    [0, 2, 1, 0, 0, 0, 0],
+    [0, 0, 2, 1, 1, 0, 0],
+    [0, 0, 0, 2, 1, 0, 0],
+    [0, 0, 0, 0, 2, 1, 1],
+    [0, 0, 0, 0, 0, 0, 1]])
+    """
+    from sympy import Dict, Dummy, SparseMatrix
+    try:
+        if len(args) not in (1, 2, 3):
+            raise TypeError
+        if not isinstance(args[-1], (dict, Dict)):
+            raise TypeError
+        if len(args) == 1:
+            rows = kwargs.get('rows', None)
+            cols = kwargs.get('cols', None)
+        elif len(args) == 2:
+            rows = cols = as_int(args[0])
+        else:
+            rows, cols = map(as_int, args[:2])
+        # fails with ValueError if any keys are not ints
+        _ = all(as_int(k) for k in args[-1])
+    except (ValueError, TypeError):
+        raise TypeError(filldedent(
+            '''unrecognized input to banded:
+            expecting [[row,] col,] {int: value}'''))
```

Comment:
Is there a way we could also allow the kwargs `rows`, `cols`, and `shape`?

Reply:
rows and cols are already supported: `banded({0: 1}, cols=3) == eye(3)`. I now note this in the docstring and improved handling for cases when either is None.

Reply:
`shape` is not yet supported in matrices as a keyword so I didn't try to add it here.